query.go

```diff
@@ -576,29 +576,179 @@ func (s *ChainService) putFilterToCache(blockHash *chainhash.Hash,
+	if !ok {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	// We're only interested in "cfilter" messages.
+	response, ok := resp.(*wire.MsgCFilter)
+	if !ok {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	// If the request filter type doesn't match the type we were expecting,
+	// ignore this message.
+	if q.filterType != request.FilterType {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	// If the response filter type doesn't match what we were expecting,
+	// ignore this message.
+	if q.filterType != response.FilterType {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	// If this filter is for a block not in our index, we can ignore it, as
+	// we either already got it, or it is out of our queried range.
+	i, ok := q.headerIndex[response.BlockHash]
+	if !ok {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	filter, err := gcs.FromNBytes(
+		builder.DefaultP, builder.DefaultM, response.Data,
+	)
+	if err != nil {
+		// Malformed filter data. We can ignore this message.
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	// Now that we have a proper filter, ensure that re-calculating the
+	// filter header hash for the header _after_ the filter in the chain
+	// checks out. If not, we can ignore this response.
+	var (
+		curHeader  = q.filterHeaders[i]
+		prevHeader = q.filterHeaders[i-1]
+	)
+	filterHeader, err := builder.MakeHeaderForFilter(filter, prevHeader)
+	if err != nil {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	if filterHeader != curHeader {
+		return query.Progress{
+			Finished:   false,
+			Progressed: false,
+		}
+	}
+
+	// At this point the filter matches what we know about it, and we
+	// declare it sane. If this is the filter requested initially, then
+	// store it for later.
+	if response.BlockHash == q.targetHash {
+		q.targetFilter = filter
+	}
+
+	// Put the filter in the cache and persistToDisk if the caller requested
+	// it.
+	// TODO(halseth): for an LRU we could take care to insert the next
+	//  height filter last.
+	dbFilterType := filterdb.RegularFilter
+	evict, err := q.cs.putFilterToCache(
+		&response.BlockHash, dbFilterType, filter,
+	)
+	if err != nil {
+		log.Warnf("Couldn't write filter to cache: %v", err)
+	}
+
+	// TODO(halseth): dynamically increase/decrease the batch size to match
+	//  our cache capacity.
+	numFilters := q.stopHeight - q.startHeight + 1
+	if evict && q.cs.FilterCache.Len() < int(numFilters) {
+		log.Debugf("Items evicted from the cache with less than %d "+
+			"elements. Consider increasing the cache size...",
+			numFilters)
+	}
+
+	if q.cs.persistToDisk {
+		err = q.cs.FilterDB.PutFilter(
+			&response.BlockHash, filter, dbFilterType,
+		)
+		if err != nil {
+			log.Warnf("Couldn't write filter to filterDB: %v", err)
```

Comment:
yeah good point!

The reason I'm hesitant to change this is cause the DB write actually _really_ slows down the process of fetching the filters from the network and so in the next PR, I basically completely divorce these two things so that the work manager doesnt get slowed down by the db writes at all

Reply:
Ah okay. Fine with me to just ignore it until the follow-up PRs. But I think in general error handling in these async calls could be something to think about.